File: src/lava/magma/runtime/message_infrastructure/test/test_multiprocessing.py
# Copyright (C) 2022 Intel Corporation
# SPDX-License-Identifier: BSD-3-Clause
# See: https://spdx.org/licenses/
import traceback
import unittest
import psutil
import os
import time
import numpy as np
from functools import partial
from enum import Enum

from message_infrastructure import Actor
from message_infrastructure import ActorStatus
from message_infrastructure.multiprocessing import MultiProcessing

import time


def nbytes_cal(shape, dtype):
    return np.prod(shape) * np.dtype(dtype).itemsize


class Builder():
    def build(self, i):
        time.sleep(0.0001)


def target_fn(*args, **kwargs):
    """
    Function to build and attach a system process to

    :param args: List Parameters to be passed onto the process
    :param kwargs: Dict Parameters to be passed onto the process
    :return: None
    """
    try:
        actor = args[0]
        builder = kwargs.pop("builder")
        idx = kwargs.pop("idx")
        builder.build(idx)
        return 0
    except Exception as e:
        print("Encountered Fatal Exception: " + str(e))
        print("Traceback: ")
        print(traceback.format_exc())
        raise e

def multiprocessing_spawn(builder, target_fn):
        """
        Spawns an actor.
        """
        mp = MultiProcessing()
        mp.start()
        #builder = Builder()
        # Build 5 actors
        for i in range(5):
            bound_target_fn = partial(target_fn, idx=i)
            mp.build_actor(bound_target_fn, builder)

        return mp, mp.actors
class TestMultiprocessing(unittest.TestCase):
    
    @unittest.skip
    def test_multiprocessing_spawn(self):
        """
        Spawns an actor.
        Checks that an actor is spawned successfully.
        """
        self.builder = Builder()

        self.mp, self.actor_list = multiprocessing_spawn(self.builder, target_fn)
        
        time.sleep(5)

        for actor in self.actor_list:
            print("status_run: ", actor.get_status())
            #self.assertEqual(actor.get_status(), 0)

        self.mp.stop(True)

    #@unittest.skip
    def test_multiprocessing_pause(self):
        builder = Builder()
        self.mp, self.actor_list = multiprocessing_spawn(builder, target_fn)
        
        time.sleep(5)
        #self.mp.pause()

        for actor in self.actor_list:
            actor.pause()
            time.sleep(1)
            print("status_pause: ", actor.get_status())
            #self.assertEqual(actor.get_status(), 1)

        self.mp.stop(True)

    @unittest.skip 
    def test_multiprocessing_shutdown(self):
        """
        Spawns an actor and sends a stop signal.
        Checks that actor is stopped successfully.
        """

        self.builder = Builder()
        self.mp, self.actor_list = multiprocessing_spawn(self.builder, target_fn)
        
        time.sleep(5)
        
        for actor in self.actor_list:
            actor.stop()
            time.sleep(1)
            print("status_stop: ", actor.get_status())
            #self.assertEqual(actor.get_status(), 2)

        self.mp.stop(True)

        # actor_list = self.mp.actor_pids
        # self.mp.stop()

        # Check that all actor PIDs no longer exist
        # for actor_pid in actor_list:
        #     self.assertFalse(psutil.pid_exists(actor_pid))
   
    @unittest.skip
    def test_get_actor_list(self):
        """
        Gets list of actors
        Checks that all actors are of Actor type
        """

        self.builder = Builder()
        self.mp, self.actor_list = multiprocessing_spawn(self.builder, target_fn)
        time.sleep(5)
        for actor in self.actor_list:
            self.assertIsInstance(actor, Actor)

        self.mp.stop(True)

def test_multiprocessing():
    mp = MultiProcessing()
    mp.start()
    builder = Builder()
    for i in range(5):
        bound_target_fn = partial(target_fn, idx=i)
        ret = mp.build_actor(bound_target_fn, builder)
        #print(ret)
    time.sleep(5)
    # shmm = mp.smm
    # for i in range(5):
    #     print("shared memory id: ", shmm.alloc_mem(8))

    actors = mp.actors
    for actor in actors:
        actor = actors[0]
        print("actor status0: ", actor.get_status())
        time.sleep(5)
        for actor in actors:
            actor.stop()
            print("actor status1: ", actor.get_status())
        mp.stop(True)

    # print("stop num: ", shmm.stop())
    # print("stop num: ", shmm.stop())

# Run unit tests
if __name__ == '__main__':
    #test_multiprocessing()
    unittest.main()

File: src/lava/magma/runtime/message_infrastructure/test/test_multiprocessing.cc
// Copyright (C) 2021 Intel Corporation
// SPDX-License-Identifier: BSD-3-Clause
// See: https://spdx.org/licenses/

#include <iostream>

#include <gtest/gtest.h>
#include <multiprocessing.h>

TEST(HelloTest, BasicAssertions) {
  // Expect two strings not to be equal.
  EXPECT_STRNE("hello", "world");
  // Expect equality.
  EXPECT_EQ(7 * 6, 42);
}

using namespace message_infrastructure;

class Builder {
  public:
    void Build(int i);
};

void Builder::Build(int i) {
  std::cout << "Builder running build " << i << std::endl;
  std::cout << "Build " << i << "... Sleeping for 10s" << std::endl;
  sleep(10);
  std::cout << "Build " << i << "... Builder complete" << std::endl;
}

AbstractActor::TargetFn TargetFunction(Builder builder, int idx) {
  std::cout << "Target Function running" << std::endl;
  builder.Build(idx);
}

TEST(TestMultiprocessing, MultiprocessingSpawn) {
  // Spawns an actor
  // Checks that actor is spawned successfully
  Builder *builder = new Builder();
  MultiProcessing mp;

  for (int i = 0; i < 5; i++) {
    std::cout << "Loop " << i << std::endl;
    auto bound_fn = std::bind(TargetFunction, std::placeholders::_1, i);
    mp.BuildActor(bound_fn(*builder));
  }
}

TEST(TestMultiprocessing, MultiprocessingShutdown) {
  // Spawns an actor and sends a stop signal
  // Checks that actor is stopped successfully
  GTEST_SKIP() << "Skipping MultiprocessingShutdown";
}

TEST(TestMultiprocessing, ActorForceStop) {
  // Force stops all running actors
  // Checks that actor status returns 1 (StatusStopped)
  GTEST_SKIP() << "Skipping ActorForceStop";
  MultiProcessing mp;
  std::vector<AbstractActor::ActorPtr>& actorList = mp.GetActors();
  for (auto actor : actorList){
    actor->ForceStop();
    int actorStatus = actor->GetStatus();
    EXPECT_EQ(actorStatus, 1);
  }
}

TEST(TestMultiprocessing, ActorRunning) {
  // Checks that acto status returns 0 (StatusRuning)
  // std::vector<ActorPtr>& actor_list = *Multiprocessing()
  MultiProcessing mp;
  std::vector<AbstractActor::ActorPtr>& actorList = mp.GetActors();
  for (auto actor : actorList){
    int actorStatus = actor->GetStatus();
    EXPECT_EQ(actorStatus, 0);
  }
}
